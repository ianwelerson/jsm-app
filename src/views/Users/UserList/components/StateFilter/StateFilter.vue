<script setup lang="ts">
import { computed, onBeforeMount, reactive, watch } from 'vue'
import { useFetch } from '@/composable/useFetch'
import type { CheckboxFormat } from '@/types'

const SHORT_COUNTRY_LIST_NUM = 5

interface StateFilterState {
  shortStateList: boolean
  countryStates: CheckboxFormat[]
  selectedState: string[]
}

const { execute: getStateList } = useFetch('/states')

const state = reactive<StateFilterState>({
  shortStateList: true,
  countryStates: [],
  selectedState: [],
})

const countryStateList = computed(() => {
  return state.shortStateList
    ? state.countryStates.slice(0, SHORT_COUNTRY_LIST_NUM)
    : state.countryStates
})

watch(
  () => state.selectedState,
  (list: string[]) => {
    emit('update', list)
  }
)

const toggleVisibleStates = () => {
  state.shortStateList = !state.shortStateList
}

const updateStatesList = async () => {
  const { data } = await getStateList()

  state.countryStates = (data.value as string[]).map(
    (state): CheckboxFormat => {
      return {
        name: state,
        key: state,
      }
    }
  )
}

onBeforeMount(async () => {
  updateStatesList()
})

const emit = defineEmits<{
  (e: 'update', list: string[]): void
}>()
</script>

<template>
  <div class="state-filter">
    <p class="state-filter__title">Por Estado</p>
    <ul class="state-filter__list">
      <li
        v-for="stateData in countryStateList"
        :key="stateData.key"
        class="state-filter__item"
      >
        <input
          data-testid="state-filter-input"
          type="checkbox"
          :name="stateData.key"
          :value="stateData.key"
          :id="`${stateData.key}-checkbox`"
          v-model="state.selectedState"
          class="state-filter__input"
        />
        <label :for="`${stateData.key}-checkbox`" class="state-filter__label">{{
          stateData.name
        }}</label>
      </li>
    </ul>
    <button
      v-if="state.countryStates.length > SHORT_COUNTRY_LIST_NUM"
      class="state-filter__show-more"
      @click="toggleVisibleStates"
      data-testid="state-filter-show-more"
    >
      {{ state.shortStateList ? 'Ver todos' : 'Ver menos' }}
    </button>
  </div>
</template>

<style lang="scss" scoped>
.state-filter {
  width: 100%;
  height: 100%;

  &__title {
    font-size: $text-lg;
    font-weight: $font-medium;
    line-height: $leading-2;
  }

  &__list {
    list-style: none;
    padding: 0;
    margin: $spacing-3 0 0 0;
  }

  &__item {
    display: flex;

    &:not(:first-child) {
      margin-top: $spacing-1;
    }
  }

  &__input {
    transform: scale(1.2);
  }

  &__label {
    font-size: $text-base;
    line-height: $leading-4;
    margin-left: $spacing-2;
    text-transform: capitalize;
  }

  &__show-more {
    background-color: $transparent;
    border: none;
    cursor: pointer;
    font-size: $text-base;
    line-height: $leading-4;
    margin-top: $spacing-1;
    padding: 0;
    text-decoration: underline;
  }
}
</style>
