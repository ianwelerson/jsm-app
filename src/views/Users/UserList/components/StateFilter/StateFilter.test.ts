import { render, fireEvent, waitFor } from '@testing-library/vue'
import StateFilter from '@/views/Users/UserList/components/StateFilter/StateFilter.vue'

vi.mock('@/composable/useFetch', () => ({
  useFetch: vi.fn().mockReturnValue({
    execute: vi.fn().mockResolvedValue({
      data: {
        value: [
          'acre',
          'alagoas',
          'amapá',
          'amazonas',
          'bahia',
          'ceará',
          'distrito federal',
          'espírito santo',
          'goiás',
        ],
      },
    }),
  }),
}))

describe('StateFilter', () => {
  it('should emit an update with the list of selected elements', async () => {
    const { getAllByTestId, emitted } = render(StateFilter)

    await waitFor(() => {
      const stateFilter = getAllByTestId('state-filter-input')

      fireEvent.click(stateFilter[0])
    })

    await waitFor(() => {
      expect(emitted('update')[0]).toEqual([['acre']])
    })
  })

  it('should show more items when user clicks on show more button', async () => {
    const { getByTestId, getByText } = render(StateFilter)

    await waitFor(() => {
      getByText('Ver todos')
    })

    const stateFilter = getByTestId('state-filter-show-more')
    fireEvent.click(stateFilter)

    await waitFor(() => {
      getByText('acre')
      getByText('Ver menos')
    })
  })

  it('should match the snapshot', async () => {
    const { html } = render(StateFilter)

    expect(html()).toMatchSnapshot()
  })
})
