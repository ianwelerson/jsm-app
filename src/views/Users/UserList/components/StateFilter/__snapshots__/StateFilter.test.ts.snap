// Vitest Snapshot v1

exports[`StateFilter > should match the snapshot 1`] = `
"<div class=\\"state-filter\\" data-v-d19c9cda=\\"\\">
  <p class=\\"state-filter__title\\" data-v-d19c9cda=\\"\\">Por Estado</p>
  <ul class=\\"state-filter__list\\" data-v-d19c9cda=\\"\\"></ul>
  <!--v-if-->
</div>"
`;
